File: android/core/src/main/kotlin/com/negfeed/epyc/core/DrawingGeometry.kt
package com.negfeed.epyc.core

import kotlin.math.ceil

/**
 * Pure geometry/timeline math ported from the legacy drawing components and
 * asserted against `tools/golden/drawing_vectors.json`.
 */
object DrawingGeometry {

    const val MAX_GAP_MS = 1000.0          // MAXIMUM_EVENT_TIME_DIFFERENCE_IN_MILLIS
    const val REPLAY_PERIOD_MS = 33.0      // REPLAY_PERIOD_IN_MILLIS
    const val MIN_PROCESSING_DISTANCE_PX = 6.0
    const val DRAW_STROKE_FRACTION = 0.01
    const val ERASE_STROKE_FRACTION = 0.03
    const val PROGRESS_BAR_HEIGHT_FRACTION = 0.03

    data class Point(val x: Double, val y: Double)
    data class ReplayTick(val period: Int, val elapsedMs: Double, val drawnCount: Int)
    data class BezierSegment(val c1: Point, val c2: Point, val end: Point)

    fun normalize(p: Point, side: Double) = Point(p.x / side, p.y / side)
    fun denormalize(p: Point, side: Double) = Point(p.x * side, p.y * side)

    fun isBeyondMinProcessingDistance(last: Point, current: Point): Boolean {
        val dx = last.x - current.x
        val dy = last.y - current.y
        return (MIN_PROCESSING_DISTANCE_PX * MIN_PROCESSING_DISTANCE_PX) < (dx * dx + dy * dy)
    }

    fun drawStrokeWidth(side: Double) = DRAW_STROKE_FRACTION * side
    fun eraseStrokeWidth(side: Double) = ERASE_STROKE_FRACTION * side

    /** First event => 0; subsequent gaps capped at MAX_GAP_MS, then accumulated. */
    fun normalizedTimestamps(timestamps: List<Double>): List<Double> {
        if (timestamps.isEmpty()) return emptyList()
        val out = DoubleArray(timestamps.size)
        out[0] = 0.0
        for (i in 1 until timestamps.size) {
            var d = timestamps[i] - timestamps[i - 1]
            if (d > MAX_GAP_MS) d = MAX_GAP_MS
            out[i] = out[i - 1] + d
        }
        return out.toList()
    }

    /** Cumulative drawn-count per 33 ms tick (touchCount == 0 path), inclusive of final period. */
    fun replayTicks(normalizedTimestamps: List<Double>): List<ReplayTick> {
        val last = normalizedTimestamps.lastOrNull() ?: 0.0
        val totalPeriods = ceil(last / REPLAY_PERIOD_MS).toInt()
        val ticks = ArrayList<ReplayTick>(totalPeriods + 1)
        var drawingIndex = 0
        var counter = 0
        while (counter <= totalPeriods) {
            val cur = counter * REPLAY_PERIOD_MS
            while (drawingIndex < normalizedTimestamps.size && normalizedTimestamps[drawingIndex] <= cur) drawingIndex++
            ticks.add(ReplayTick(counter, cur, drawingIndex))
            counter++
        }
        return ticks
    }

    /** Catmull-Rom spline → cubic Bezier segments (Paper.js `path.smooth()` equivalent). */
    fun catmullRomToBezier(points: List<Point>): List<BezierSegment> {
        if (points.size < 2) return emptyList()
        val segments = ArrayList<BezierSegment>(points.size - 1)
        for (i in 0 until points.size - 1) {
            val p0 = points[maxOf(i - 1, 0)]
            val p1 = points[i]
            val p2 = points[i + 1]
            val p3 = points[minOf(i + 2, points.size - 1)]
            val c1 = Point(p1.x + (p2.x - p0.x) / 6.0, p1.y + (p2.y - p0.y) / 6.0)
            val c2 = Point(p2.x - (p3.x - p1.x) / 6.0, p2.y - (p3.y - p1.y) / 6.0)
            segments.add(BezierSegment(c1, c2, p2))
        }
        return segments
    }
}

File: ios/EpycCore/Sources/EpycCore/DrawingGeometry.swift
import Foundation

/// Pure geometry/timeline math ported from the legacy drawing components:
///   - `src/components/drawing-canvas/drawing-canvas.ts`        (normalize, stroke widths)
///   - `src/components/recording-drawing-canvas/...`            (min processing distance)
///   - `src/components/replaying-drawing-canvas/...`            (timestamp normalize + replay ticks)
/// Asserted against `tools/golden/drawing_vectors.json`.
public enum DrawingGeometry {

    // Constants from the legacy components.
    public static let maxGapMs: Double = 1000      // MAXIMUM_EVENT_TIME_DIFFERENCE_IN_MILLIS
    public static let replayPeriodMs: Double = 33  // REPLAY_PERIOD_IN_MILLIS
    public static let minProcessingDistancePx: Double = 6
    public static let drawStrokeFraction: Double = 0.01   // strokeWidth = 0.01 * side
    public static let eraseStrokeFraction: Double = 0.03
    public static let progressBarHeightFraction: Double = 0.03

    public struct Point: Equatable, Sendable { public var x: Double; public var y: Double
        public init(x: Double, y: Double) { self.x = x; self.y = y } }

    // MARK: Coordinate normalization

    public static func normalize(_ p: Point, side: Double) -> Point {
        Point(x: p.x / side, y: p.y / side)
    }
    public static func denormalize(_ p: Point, side: Double) -> Point {
        Point(x: p.x * side, y: p.y * side)
    }

    // MARK: Recording — min processing distance (squared comparison, like the TS)

    public static func isBeyondMinProcessingDistance(last: Point, current: Point) -> Bool {
        let dx = last.x - current.x
        let dy = last.y - current.y
        return (minProcessingDistancePx * minProcessingDistancePx) < (dx * dx + dy * dy)
    }

    public static func drawStrokeWidth(side: Double) -> Double { drawStrokeFraction * side }
    public static func eraseStrokeWidth(side: Double) -> Double { eraseStrokeFraction * side }

    // MARK: Replay — timestamp normalization

    /// First event => 0; subsequent gaps capped at `maxGapMs`, then accumulated.
    public static func normalizedTimestamps(_ timestamps: [Double]) -> [Double] {
        guard !timestamps.isEmpty else { return [] }
        var out = [Double](repeating: 0, count: timestamps.count)
        out[0] = 0
        for i in 1..<timestamps.count {
            var d = timestamps[i] - timestamps[i - 1]
            if d > maxGapMs { d = maxGapMs }
            out[i] = out[i - 1] + d
        }
        return out
    }

    public struct ReplayTick: Equatable, Sendable {
        public var period: Int
        public var elapsedMs: Double
        public var drawnCount: Int
    }

    /// Drives the 33 ms replay loop (touchCount == 0 path): at each tick, every event
    /// with normalized timestamp <= elapsed has been drawn. Returns the cumulative
    /// drawn-count per tick from period 0 through the final period (inclusive).
    public static func replayTicks(normalizedTimestamps norm: [Double]) -> [ReplayTick] {
        let last = norm.last ?? 0
        let totalPeriods = Int(ceil(last / replayPeriodMs))
        var ticks: [ReplayTick] = []
        var drawingIndex = 0
        var counter = 0
        while counter <= totalPeriods {
            let cur = Double(counter) * replayPeriodMs
            while drawingIndex < norm.count && norm[drawingIndex] <= cur { drawingIndex += 1 }
            ticks.append(ReplayTick(period: counter, elapsedMs: cur, drawnCount: drawingIndex))
            counter += 1
        }
        return ticks
    }

    // MARK: Catmull-Rom smoothing (visual parity with Paper.js `path.smooth()`)

    /// Converts a Catmull-Rom spline through `points` into cubic Bezier control
    /// points, the standard equivalent of Paper.js's default smoothing. Returned as
    /// (control1, control2, endPoint) segments to feed a Bezier path renderer.
    public static func catmullRomToBezier(_ points: [Point]) -> [(c1: Point, c2: Point, end: Point)] {
        guard points.count >= 2 else { return [] }
        var segments: [(c1: Point, c2: Point, end: Point)] = []
        for i in 0..<(points.count - 1) {
            let p0 = points[max(i - 1, 0)]
            let p1 = points[i]
            let p2 = points[i + 1]
            let p3 = points[min(i + 2, points.count - 1)]
            let c1 = Point(x: p1.x + (p2.x - p0.x) / 6.0, y: p1.y + (p2.y - p0.y) / 6.0)
            let c2 = Point(x: p2.x - (p3.x - p1.x) / 6.0, y: p2.y - (p3.y - p1.y) / 6.0)
            segments.append((c1: c1, c2: c2, end: p2))
        }
        return segments
    }
}
